Full text improvements

oskardudycz opened the pull request:
# Description
Currently it's Full Text search allow to define only one index per document. For most of the simple cases it's enough, but it might be not efficient to index whole `data`, when only one of few fields should be filtered. It's also not enough if we'd have to support more than language and filter only texts in specific language.

## Type of change

- added possibility to define custom data config (via config parameter of AddFullTextIndex method or via selected document members),
- changed index name to contain regional config - thus allow to configure more than one FTS for different languages,
- added full text index attribute to make easier fts fields definiton,
- added support for Conjoined Tenancy.

Also added possibility to define custom index name in AddUniqueIndex method with members expression.

 [x] New feature (non-breaking change)

# How Has This Been Tested?

Added new [tests to already existing for FTS](https://github.com/JasperFx/marten/pull/1167/files#diff-3cd0582ff65551d6588c90142e70e847)

Manually checked also execution plans for:

1. Two FTS indexes for different languages and single members selected
- Table definition
```
CREATE TABLE public.mt_doc_user
(
    id uuid NOT NULL,
    data jsonb NOT NULL,
    mt_last_modified timestamp with time zone DEFAULT transaction_timestamp(),
    mt_version uuid NOT NULL DEFAULT (md5(((random())::text || (clock_timestamp())::text)))::uuid,
    mt_dotnet_type character varying COLLATE pg_catalog."default",
    CONSTRAINT pk_mt_doc_user PRIMARY KEY (id)
)
WITH (
    OIDS = FALSE
)
TABLESPACE pg_default;

CREATE INDEX mt_doc_user_french_idx_fts
    ON public.mt_doc_user USING gin
    (to_tsvector('french'::regconfig, data ->> 'LastName'::text))
    TABLESPACE pg_default;

CREATE INDEX mt_doc_user_italian_idx_fts
    ON public.mt_doc_user USING gin
    (to_tsvector('italian'::regconfig, data ->> 'FirstName'::text))
    TABLESPACE pg_default;
```
- query plan:

![obraz](https://user-images.githubusercontent.com/5562528/51089359-1a7fca00-176c-11e9-9034-b3155b9250bf.png)

2. Single FTS index for default language and two members selected
- Table definition
```
CREATE TABLE public.mt_doc_user
(
    id uuid NOT NULL,
    data jsonb NOT NULL,
    mt_last_modified timestamp with time zone DEFAULT transaction_timestamp(),
    mt_version uuid NOT NULL DEFAULT (md5(((random())::text || (clock_timestamp())::text)))::uuid,
    mt_dotnet_type character varying COLLATE pg_catalog."default",
    CONSTRAINT pk_mt_doc_user PRIMARY KEY (id)
)
WITH (
    OIDS = FALSE
)
TABLESPACE pg_default;

CREATE INDEX mt_doc_user_english_idx_fts
    ON public.mt_doc_user USING gin
    (to_tsvector('english'::regconfig, ((data ->> 'FirstName'::text) || ' '::text) || (data ->> 'LastName'::text)))
    TABLESPACE pg_default;
```
- query plan:

![obraz](https://user-images.githubusercontent.com/5562528/51089394-9a0d9900-176c-11e9-9602-ac5eca1129dc.png)

@CodingGorilla: I would love to see your point of view about those changes as you're inital author of this feature.

I plan to also update documentation, when we have agreement on the API.

CodingGorilla commented:
It all looks fine to me.  I am by no means an expert on PostgresSQL or full text indexing in it.  I wrote the original feature as a bit of a learning experience with PGSQL.  These changes actually look pretty minimal, so far.

oskardudycz commented:
@CodingGorilla thank you for your feedback 👍 Yes, it's just step forward, I'm also not such expert about FTS on Postgres. So for sure it's just the first step, not the last one.

Looking at the query plans it seems that it's using the defined indexes, so from this stand point - it looks good. For sure we'd need some more investigation on performance, maybe some benchmarks etc.

oskardudycz commented:
Guys, I added first round of samples and documentation as @jokokko and @CodingGorilla were fine with initial concept. I'll try to finalize rest of documentation and testing tomorrow or on wednesday.

oskardudycz commented:
@jeremydmiller @CodingGorilla @mysticmind @jokokko I added possibilty to extend `MartenQueryable` with `WhereFragment` and used it in `FullTextSearch`. See more in this commit: https://github.com/JasperFx/marten/pull/1167/commits/8a79740d8faf2e403acaa25891cd20db6b2ad7c3

This makes possible to join FullTextSearch mechanism together with queryable. It'll make also possible to use it even with any other custom where.

I know that this change might be controvercial, but imho it'll make in future easy extensible plugging in other such queries. It will also make Marten users to even write their own custom wheres without needing to extend Marten core logic.

I'd be grateful for the feedback!

CodingGorilla commented:
@oskardudycz Can you show an example of how you expect this to be used?  It seems like there's likely some confusion as to how this might end up working that could cause queries to become 'garbeled' (for lack of a better way to explain it).

oskardudycz commented:
@CodingGorilla I added also sample in code for the usage of conjoined FullText search with queryable. I added also FullTextSearch methods to QueryableExtensions to allow usage on IQueryable<T> interface so:

```
var travelBlogPosts = session.Query<BlogPost>()
                    .Where(blogPost => blogPost.Category == "Travel")
                    .Search("Lorem ipsum")
                    .ToList()
```

CodingGorilla commented:
Ok, never mind my original question, my brain was still stuck in relational query mode. :)  But, while this looks like it would work, it seems like it kind of defeats the purpose of a full text search doesn't it?  You're doing a FTS inside one or more fields?

oskardudycz commented:
@CodingGorilla well, with this improvements you have now just more options than you had. You can do search as was originally designed - so on all contents of data field, but if you'd like to you can select to index only one (or more) fields from data - so reuse documents.

It also gives possibility to define more thant one FTS index on the document - so support eg. translations. You can have document with different versions for different languages and search with exact language (the Blog Post or Article could be good example of that).

Also filtering out some records (so joining FTS searches with query) imho is one of the useful features, even from performance perspective.

If you'll be doing fts then it will try to find and reuse defined index or filter whole data (see the logic for that [here](https://github.com/JasperFx/marten/pull/1167/files#diff-6019345c44b68252abc72ff417a6f4f0R280)). Query plans confirms that.

Of course normally I'd suggest to have specific document designed for FTS, but purpose of this PR is just to give users more options and flexibility.

CodingGorilla commented:
I'm not always a big fan of giving people more ways to accomplish the same thing.  It adds complexity and makes troubleshooting and support very difficult at times.  But if this is a feature that you find useful, then I say "Make it so!". 😆

oskardudycz commented:
@CodingGorilla yes, I know what you're saying. That's why I wrote that it's a bit controvercial ;) I already got such feature request from my colleagues. I also personally see the value on that, but I'm also a bit afraid that people my overuse this stuff. I'll try to put some guidance/best practices in the docs.

@CodingGorilla Thank you a lot for a good constructive feedback! And what's more - I learned new word - `garbeled` 😆

@jeremydmiller @mysticmind @jokokko what are your thoughts?

cocowalla commented:
> Also filtering out some records (so joining FTS searches with query) imho is one of the useful features, even from performance perspective

As a (happy!) user of Marten, I would find this really useful - I would think it's a really common scenario too.

oskardudycz commented:
Thanks @cocowalla, appreciate the feedback 👌

jeremydmiller commented:
@oskardudycz Oh man, I should have been paying more attention. I want to push you toward a little different implementation in the Linq mechanics. I think that explicitly adding `IWhereFragment` objects to `IMartenQueryable` is polluting the internal APIs and forcing you to make a lot of new object allocations from the before and adding complexity I don't think we want.

What I would suggest is using the `IMethodCallParser` mechanism instead, and let the normal `IMartenQueryable` mechanics do its thing. So instead of these methods directly on `IMartenQ
ueryable` adding WhereFragments directly:

```
        /// <summary>
        /// Performs a full text search against <typeparamref name="TDoc"/>
        /// </summary>
        /// <param name="searchTerm">The text to search for.  May contain lexeme patterns used by PostgreSQL for full text searching</param>
        /// <param name="regConfig">The dictionary config passed to the 'to_tsquery' function, must match the config parameter used by <seealso cref="DocumentMapping.AddFullTextIndex(string)"/></param>
        /// <remarks>
        /// See: https://www.postgresql.org/docs/10/static/textsearch-controls.html#TEXTSEARCH-PARSING-QUERIES
        /// </remarks>
        IMartenQueryable<T> Search(string searchTerm, string regConfig = FullTextIndex.DefaultRegConfig);

        /// <summary>
        /// Performs a full text search against <typeparamref name="TDoc"/> using the 'plainto_tsquery' search function
        /// </summary>
        /// <param name="queryText">The text to search for.  May contain lexeme patterns used by PostgreSQL for full text searching</param>
        /// <param name="regConfig">The dictionary config passed to the 'to_tsquery' function, must match the config parameter used by <seealso cref="DocumentMapping.AddFullTextIndex(string)"/></param>
        /// <remarks>
        /// See: https://www.postgresql.org/docs/10/static/textsearch-controls.html#TEXTSEARCH-PARSING-QUERIES
        /// </remarks>
        IMartenQueryable<T> PlainTextSearch(string searchTerm, string regConfig = FullTextIndex.DefaultRegConfig);

        /// <summary>
        /// Performs a full text search against <typeparamref name="TDoc"/> using the 'phraseto_tsquery' search function
        /// </summary>
        /// <param name="queryText">The text to search for.  May contain lexeme patterns used by PostgreSQL for full text searching</param>
        /// <param name="regConfig">The dictionary config passed to the 'to_tsquery' function, must match the config parameter used by <seealso cref="DocumentMapping.AddFullTextIndex(string)"/></param>
        /// <remarks>
        /// See: https://www.postgresql.org/docs/10/static/textsearch-controls.html#TEXTSEARCH-PARSING-QUERIES
        /// </remarks>
        IMartenQueryable<T> PhraseSearch(string searchTerm, string regConfig = FullTextIndex.DefaultRegConfig);
```

Let's make these strictly extension methods like:

```
public static class FullTextExtensions
{
    public static bool PhraseSearch(this T entity, string searchTerm, string regConfig = FullTextIndex.DefaultRegConfig)
    {
        return true; // doesn't really matter here
     }
}
```

and a corresponding `IMethodCallParser` for each new method that returns the same IWhereFragment you've already built.

If you want, you can come back with this too in the same extension method class:

```
// just syntactical sugar
public static IMartenQueryable<T> PhraseSearch(this IMartenQueryable<T> queryable, args)
{
    
    return queryable.Where(x => x.PhraseSearch(args));'
}
```

By relying on the IMethodCallParser, you're making fewer changes to the internals of MartenQueryable and doing fewer object allocations

oskardudycz commented:
Ok, it's fixed. I'll polish the changes (tests, documentation) and will try to finish that in the next 2-3 days.

oskardudycz commented:
Ok, I added documentation, more tests. It's ready for final review! :)
@jeremydmiller @mysticmind @CodingGorilla @jokokko

mysticmind commented:
@oskardudycz I had merged the PR #1168 with few minor tweak, this is resulting in a conflict in this PR for README.md, I will take a look to fix it.

oskardudycz commented:
@mysticmind sure, but if you'd like I can do rebase by myself.

oskardudycz commented:
@mysticmind I did additonal rebase to force Github to not scream about conflicts ;)

mdissel commented on src/Marten/Linq/WhereFragments/FullTextWhereFragment.cs:
```diff
@@ -0,0 +1,33 @@
+        {
+            var dataConfig = GetDataConfig(mapping, regConfig);
+
+            return $"to_tsvector('{regConfig}', {dataConfig}) @@ {searchFunction}('{regConfig}', '{searchTerm}')";
```
Sql injection possible?

jokokko replied:
Good catch (and shame on me for being too hasty)!. It is. `x.Search(@"somefilter ') ; drop table mt_doc_blogpost ; select ('"))`.

oskardudycz replied:
Yeah, I forgot about that. I was supposed to fix that issue https://github.com/JasperFx/marten/issues/1170. I'll work on that

oskardudycz replied:
@mdissel @jokokko  @RobSchoenaker Fixed in https://github.com/JasperFx/marten/pull/1167/commits/b31120cbd2b91844f3905bed69be998a6bd5908a

oskardudycz commented on src/Marten/Linq/Parsing/FullTextSearchMethodCallParser.cs:
```diff
@@ -0,0 +1,55 @@
+                throw new ArgumentException("Search Term needs to be string", "searchTerm");
+
+            if (expression.Arguments.Count > 2 && expression.Arguments[2].Type != typeof(string))
+                throw new ArgumentException("Reg config needs to be string", "searchTerm");
```
Fixed.

oskardudycz commented on documentation/documentation/documents/querying/linq.md:
```diff
@@ -211,6 +211,32 @@ Query data from all tenants using `AnyTenant` method.
+
+<[sample:text_search_combined_with_other_query_sample]>
+
+They alow also to specify language (regConfig) of the text search query (by default `english` is being used)
```
Fixed.

oskardudycz commented on documentation/documentation/documents/querying/linq.md:
```diff
@@ -211,6 +211,32 @@ Query data from all tenants using `AnyTenant` method.
 
+## Text Search
+
+Postgres constains built in [Text Search functions](https://www.postgresql.org/docs/10/textsearch-controls.html). They enable the possibility to do more sophisticated searching through text fields. Marten gives possibility to define <[linkto:documentation/documents/configuration/full_text;title=Full Text Indexes]> and perform queries on them.
```
Fixed.

oskardudycz commented:
@mysticmind I added asserts for some of the sample methods in tests. Removed `[Fact]` attribute from the sample methods that are covered in other tests.

All comments addressed.
